### ISSUE TITLE
Support reactivity in ES6 classes

### ISSUE THREAD

**Issue by thiagomajesk** [salesforce/observable-membrane#42]
Hello everyone! 
A [question](https://github.com/alpinejs/alpine/issues/270) arose in the [AlpineJs](https://github.com/alpinejs) project regarding whether the usage of ES6 classes was possible or not. Since they use observable-membrane as the reactivity engine I'd like to know if this is a hard limitation on the architecture or is something reasonable to ask for. My current understanding is that right now, we can only observe object literals and not classes.

**Comment by pmdartus** [comment]
This restriction is on purpose. The main issue with observing objects that don't inherit from the `Object.constructor` or `null` is that it introduces unexpected behaviors related to the `this` value.

In the following example, we use a WeakMap to associate each `Foo` instance with its internal state. If the membrane wraps the instance, the following code would fail because the `this` value in the `logState` method is not the `Foo` instance but a `Proxy` wrapping the `Foo` instance.

```js
const internal = new WeakMap();

class Foo {
  constructor() {
  	internal.set(this, { value: 'foo' });
  }
  
  logState() {
  	const state = internal.get(this);
        console.log(state);
  }
}

const membrane = new ObservableMembrane();

const proxified = membrane.getProxy(new Foo());
proxified.logState();
```

**Comment by thiagomajesk** [comment]
@pmdartus Is there a way to access the underlining `Foo` or bind the proper `this` for this context?

**Comment by SimoTod** [comment]
Is the issue only related to the `this` when calling a function? 
If yes, wouldn't be possible to solve it using an apply trap so you can re-bind the right context?

I appreciate that, since it's a commercial product, you probably don't accept external PRs and it will take a while to get it tested and signed off but it could be a solution to explore.

Thanks for your hard work.

**Comment by caridy** [comment]
Let me try to clarify... the membrane implementation in this library does not wrap functions into proxies (that include methods and accessors). The main purpose is to observe operations via property mutation using Object Property Assignment notation, which is consider a mutation from the outside. If the mutation is from inside the object (via a method, or accessor, via `this` value), the side effects of those functions can't really be observed reliably, why bother?

Additionally, keep in mind that doing so will increase the complexity and the size of this library considerable, also considerable performance impact, e.g.: `[].push(1)` is going to have to return a proxy for the `push` function, which will try to observe something that is observable anyways via Object Property Assignment notation.

If you want to provide more details about what you're trying to do, we might be able to help, and provide some feedback.

**Comment by SimoTod** [comment]
Hi @caridy 

Properties in ES6 are not observable through the membrane by design. When asked, we were told that the reason is that `this` is not available in the proxied object so I was wondering if the apply trap was a viable way. Given that it's not an option, it's not a problem.

I know that the membrane doesn't wrap functions into proxies and doing so will cost an additional indirection for each call, but it's conceptually the same for objects, the membrane accepts a performance impact and in return it can observe mutations on standard objects.
The get trap would just return a function proxy and the proxy would just use Reflection.apply to call the function using the right context. That would remove the restriction on what is observable and what is not.

As I said, not a problem. Thanks for replying.

**Comment by caridy** [comment]
@SimoTod membranes are hard, don't underestimate that fact. Functions and constructors are a lot more complicated than just properties, and you will have a touch time determine what functions to wrap as well. Not all functions can be wrapped, (ex. all DOM apis will do brand check, which fails through the membrane). If you want to give it a try, let us know, but my only ask is to define exactly what you want to do, and what kind of functions are you planning to wrap.

**Comment by SimoTod** [comment]
I'll have a look at the whole source code in the next few days and I'll let you know if I decide to give it a go, thanks. 👍

**Comment by cliveb** [comment]
@SimoTod Perhaps to help your investigation, this ES membrane explainer by @ajvincent  https://ajvincent.github.io/es-membrane/mappings/one-to-one.html

**Comment by SimoTod** [comment]
Thanks @cliveb, I'll have a look.

@caridy 
At the minute, I've got this simple WIP (I know, the code is not nice and neat, bare with me, i've just given it a go tonight): https://github.com/salesforce/observable-membrane/compare/master...SimoTod:feature/es6-classes-support?expand=1

It makes any objects observable. Function don't use proxies (but they could, I'll explain later un this post) but, for now, they just bind the context when needed.
The if guard might need further tweaking but, at the moment:
- Array are ignored in that guard because they are kind of exceptional in JS.
- Functions which belong directly to the object don't bind the original context (so property accesses/assignments in custom function will go through the membrane).

This is a codepen using the forked version and replicating/upgrading the example in the repo: https://codepen.io/SimoTod/pen/QWbBpyM

As you can see, functions that are defined in the prototype and change the internal state, skip the membrane. Using special proxies with a membrane hook could fix it.

Basically that was my initial idea. It might be that it's not valid, I didn't have much time to test/think about that.

**Comment by caridy** [comment]
3 fundamental issues:

* every time the get is called, you get a new value back for that function, which means `foo.fn === foo.fn` returns `false`.
* it is impossible to do `foo.fn.apply` or `foo.fn.call` with a custom context value because of the bind, it does nothing.
* own methods vs inherited methods are effectively different, why? what happen if I do `obj.__proto__.foo`?

**Comment by caridy** [comment]
for the record, the more advanced membrane we have built so far is https://github.com/caridy/secure-javascript-environment, it is not about reactivity, but it is a very solid transparent membrane implementation, you can probably learn one thing or two from that repo.

**Comment by SimoTod** [comment]
Thanks I'll try to have a look when I can.
To clarify, I wasn't suggesting that those few lines were the final solution. 
In my head the function should still be wrapped in a proxy (doing the custom binding in the apply trap when needed) which can be stored in a weakmap (fixing the identity issue), the obj.fn.apply thing could probably be fixed but i would need to look into it (the right "if" guard needs a bit more research and fine tuning).
About the third point, if you currently 
define a function in your object, and the function changes a property using `this.variable = something`, the changes goes through the membrane so it was just to avoid a breaking change.
Thanks for your inputs everyone, since it's a closed issue I'll stop bugging you all.

**Comment by caridy** [comment]
keep bugging us, no worries! curious to see where does this land!

**Comment by SimoTod** [comment]
@caridy so... I haven't read the other repo yet (Sorry, it's quite late here in UK andI probably need a fresh mind to figure out what it does).

However, I've updated the repo to match my original idea (first time I use TypeScript so ignore the type-hinting please. I probably kept everything too generic :D).

This implementation doesn't break the identity principle (foo.fn === foo.fn) and allows `.call` and `.apply` to bind different context.
I've also removed the check for own methods vs inherited (it would be more complicate in the apply trap) and added a hook in the membrane. Arrays are still an exception but they probably should not be one.

Diff -> https://github.com/salesforce/observable-membrane/compare/master...SimoTod:feature/es6-classes-support?expand=1
Demo -> https://codepen.io/SimoTod/pen/QWbBpyM (I've thrown in a lot of code without an order but I'm sure you'll figure it out, all the object type so far are observable and reactive).

**Comment by caridy** [comment]
That's more realistic, few notes:

* I do not understand yet why functionCall is a new option, what does it do? why is that important? how is the observation of what that function will do tracked?
* in the apply trap, you also have to unwrap arguments
* in the apply trap, you must wrap the result of the operation
* the shadow target for a function is a little bit more complicated than that to avoid adding an observable difference (look at the other repo linked above for more details)

**Comment by SimoTod** [comment]
The new option allows me to track when a method is called (just realised that the first parameter should be the name rather than the whole function).
In the demo I posted, for instance, it allows me to make the page react to object updates when their internal status changes due to one of their methods. 

Good call about wrapping the result back in case we return this or another object which was supposed to be observable. 

Not sure about unwrapping the arguments, why should I? Do you have an example where passing a proxy would break the function call? Is it because we don't want to observe them inside the function scope for consistency with `this`? 

I should have time to have a look and try to understand the other repo tonight, I assume you mean that the shadow target should have the same arguments as the original method (not sure why, but it's out of my knowledge zone so I bet there's a reason for that).
